src/python/pants/source/source_root.py

```diff
@@ -361,6 +366,25 @@ def _do_add_pattern(self, pattern, langs, category):
     node.category = category
     node.is_terminal = True
 
+  def traverse(self) -> Set[UncanonicalizedSourceRoot]:
+    uncanonicalized_source_roots = set()
```

Comment:
Usually you'll want to type empty collection initializations like this. Otherwise MyPy can only infer `Set[Any]`.
```suggestion
    uncanonicalized_source_roots: Set[UncanonicalizedSourceRoot] = set()
```